libraries/AP_HAL/UARTDriver.cpp

```diff
@@ -226,3 +232,128 @@ void AP_HAL::UARTDriver::log_stats(const uint8_t inst, StatsTracker &stats, cons
 }
 #endif // HAL_LOGGING_ENABLED
 #endif // HAL_UART_STATS_ENABLED
+
+
+/*
+ * Support for logging all incoming/outgoing data from a serial port
+ */
+
+#if HAL_UART_DEBUG_LOGGING_ENABLED
+
+/*
+  log some data for debugging
+*/
+void AP_HAL::UARTDriver::log_read_data(const uint8_t *data, uint16_t length)
+{
+    log_data(data, length, false);
+}
+void AP_HAL::UARTDriver::log_written_data(const uint8_t *data, uint16_t length)
+{
+    log_data(data, length, true);
+}
+void AP_HAL::UARTDriver::log_data(const uint8_t *data, uint16_t length, bool is_written_data)
+{
+    // first see if we actually do want to log:
+    if ((_last_options & OPTION_LOGGING) == 0) {
+        return;
+    }
+
+    // find logging instance corresponding to this UART:
+    if (loginfo == nullptr) {
+        loginfo = NEW_NOTHROW LogInfo();
+        if (loginfo == nullptr) {
+            return;
+        }
+        loginfo->instance = instance;
+        loginfo->next = backend_loginfos;
+        backend_loginfos = loginfo;
+    }
+    if (!log_thread_created) {
+        log_thread_created = true;
+        hal.scheduler->thread_create(FUNCTOR_BIND_MEMBER(&AP_HAL::UARTDriver::logging_start, void), "uart_log", 4096, AP_HAL::Scheduler::PRIORITY_IO, 0);
+    }
+
+    // write packet into buffer:
+    struct PACKED Header {
+        enum class Flag {
+            IS_WRITTEN_DATA = 1,
+        };
+        uint32_t magic = 0x7fe53b04U;
```

Comment:
The format is different to the files written by AP_GPS so the magic should be at least a bit different. Probably just `0x7fe53b05U` is fine.